MobileBuy/buy/src/main/java/com/shopify/buy3/RealGraphCall.java

```diff
@@ -0,0 +1,106 @@
+package com.shopify.buy3;
+
+import android.support.annotation.NonNull;
+
+import com.google.gson.Gson;
+import com.shopify.graphql.support.AbstractResponse;
+import com.shopify.graphql.support.Query;
+import com.shopify.graphql.support.SchemaViolationError;
+import com.shopify.graphql.support.TopLevelResponse;
+
+import java.io.IOException;
+
+import okhttp3.Call;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+final class RealGraphCall<T extends AbstractResponse<T>> implements GraphCall<T> {
+  private static final String ACCEPT_HEADER = "application/json";
+  private static final String CONTENT_TYPE_HEADER = "application/graphql";
+  private static final MediaType GRAPHQL_MEDIA_TYPE = MediaType.parse("application/graphql; charset=utf-8");
+
+  private final Query query;
+  private final HttpUrl serverUrl;
+  private final Call.Factory httpCallFactory;
+  private final ResponseConverter<T> responseConverter;
+  private volatile Call httpCall;
+  private boolean executed;
+
+  RealGraphCall(final Query query, final HttpUrl serverUrl, final Call.Factory httpCallFactory,
+                final ResponseConverter<T> responseConverter) {
+    this.query = query;
+    this.serverUrl = serverUrl;
+    this.httpCallFactory = httpCallFactory;
+    this.responseConverter = responseConverter;
+  }
+
+  @Override public void cancel() {
+    Call call = httpCall;
+    if (call != null) {
+      call.cancel();
+    }
+  }
+
+  @SuppressWarnings("CloneDoesntCallSuperClone")
+  @NonNull @Override public GraphCall<T> clone() {
+    return new RealGraphCall<>(query, serverUrl, httpCallFactory, responseConverter);
+  }
+
+  @NonNull @Override public T execute() throws IOException, SchemaViolationError {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already Executed");
+      executed = true;
+    }
+    httpCall = httpCall();
+    Response response = httpCall.execute();
+    TopLevelResponse topLevelResponse = parse(response);
+    return responseConverter.convert(topLevelResponse);
+  }
+
+  @NonNull @Override public GraphCall<T> enqueue(@NonNull final Callback<T> callback) {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already Executed");
```

Comment:
This is precondition exception, as we make sure that we don't run queries multiple times. This logic is similar to what behind `okhttp3.RealCall#enqueue`:
```
@Override public void enqueue(Callback responseCallback) {
    synchronized (this) {
      if (executed) throw new IllegalStateException("Already Executed");
      executed = true;
    }
    captureCallStackTrace();
    client.dispatcher().enqueue(new AsyncCall(responseCallback));
  }
```